File: internal/gitinfo/worktree.go
package gitinfo

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

const maxWorktrees = 64

// WorktreeIdentity is content-free host-observed repository metadata. It does
// not imply permission to read, write, launch, or run Git in the worktree.
type WorktreeIdentity struct {
	Root         string `json:"root"`
	GitDir       string `json:"gitDir"`
	CommonGitDir string `json:"commonGitDir"`
	Branch       string `json:"branch,omitempty"`
	Head         string `json:"head"`
	Linked       bool   `json:"linked"`
}

type ExistingWorktree struct {
	Root   string `json:"root"`
	Branch string `json:"branch,omitempty"`
	Head   string `json:"head"`
}

type WorktreeBounds struct {
	Shown int `json:"shown"`
	Total int `json:"total"`
	More  int `json:"more"`
}

// InspectWorktree validates candidate against the repository containing
// projectRoot using read-only Git commands and canonical filesystem identity.
func (s Service) InspectWorktree(
	ctx context.Context,
	projectRoot string,
	candidate string,
) (WorktreeIdentity, error) {
	canonicalCandidate, err := canonicalExistingPath(candidate)
	if err != nil {
		return WorktreeIdentity{}, fmt.Errorf("canonicalize selected worktree path: %w", err)
	}
	project, err := s.inspectWorktreeIdentity(ctx, projectRoot)
	if err != nil {
		return WorktreeIdentity{}, fmt.Errorf("inspect project repository: %w", err)
	}
	selected, err := s.inspectWorktreeIdentity(ctx, candidate)
	if err != nil {
		return WorktreeIdentity{}, fmt.Errorf("inspect selected worktree: %w", err)
	}
	if project.CommonGitDir != selected.CommonGitDir {
		return WorktreeIdentity{}, errors.New("selected worktree belongs to a different repository")
	}
	if !pathWithinRoot(selected.Root, canonicalCandidate) {
		return WorktreeIdentity{}, errors.New("selected path is outside the inspected worktree")
	}
	runner := s.Runner
	if runner == nil {
		runner = ExecRunner{}
	}
	listedOutput, err := runner.Output(ctx, projectRoot, "worktree", "list", "--porcelain", "-z")
	if err != nil {
		return WorktreeIdentity{}, fmt.Errorf("list repository worktrees: %w", err)
	}
	listed, _, err := parseWorktreeList(listedOutput)
	if err != nil {
		return WorktreeIdentity{}, fmt.Errorf("parse repository worktrees: %w", err)
	}
	member := false
	for _, worktree := range listed {
		if worktree.Root == selected.Root {
			member = true
			break
		}
	}
	if !member {
		return WorktreeIdentity{}, errors.New("selected worktree is not registered with the project repository")
	}
	return selected, nil
}

func pathWithinRoot(root, candidate string) bool {
	relative, err := filepath.Rel(filepath.Clean(root), filepath.Clean(candidate))
	return err == nil && relative != ".." &&
		!strings.HasPrefix(relative, ".."+string(filepath.Separator))
}

func (s Service) inspectWorktreeIdentity(
	ctx context.Context,
	root string,
) (WorktreeIdentity, error) {
	runner := s.Runner
	if runner == nil {
		runner = ExecRunner{}
	}
	output, err := runner.Output(ctx, root,
		"rev-parse", "--path-format=absolute", "--is-inside-work-tree",
		"--show-toplevel", "--absolute-git-dir", "--git-common-dir",
		"--is-bare-repository", "--verify", "HEAD",
	)
	if err != nil {
		return WorktreeIdentity{}, err
	}
	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	if len(lines) != 6 {
		return WorktreeIdentity{}, errors.New("malformed worktree identity")
	}
	inside, insideErr := strconv.ParseBool(lines[0])
	bare, bareErr := strconv.ParseBool(lines[4])
	if insideErr != nil || bareErr != nil || !inside || bare {
		return WorktreeIdentity{}, errors.New("selected path is not a non-bare worktree")
	}
	canonical := make([]string, 3)
	for index, path := range lines[1:4] {
		canonical[index], err = canonicalExistingPath(path)
		if err != nil {
			return WorktreeIdentity{}, err
		}
	}
	for _, index := range []int{0, 1, 2} {
		info, statErr := os.Stat(canonical[index])
		if statErr != nil || !info.IsDir() {
			return WorktreeIdentity{}, errors.New("worktree identity path is not a directory")
		}
	}
	branchOutput, branchErr := runner.Output(ctx, canonical[0],
		"symbolic-ref", "--quiet", "--short", "HEAD",
	)
	if branchErr != nil && !errors.Is(branchErr, ErrCommandFailed) {
		return WorktreeIdentity{}, branchErr
	}
	branch := strings.TrimSpace(string(branchOutput))
	if strings.ContainsAny(branch, "\x00\r\n") || len(branch) > 512 {
		return WorktreeIdentity{}, errors.New("malformed worktree branch")
	}
	head := strings.TrimSpace(lines[5])
	if !validObjectID(head) {
		return WorktreeIdentity{}, errors.New("malformed worktree HEAD")
	}
	return WorktreeIdentity{
		Root: canonical[0], GitDir: canonical[1], CommonGitDir: canonical[2],
		Branch: branch, Head: strings.ToLower(head),
		Linked: canonical[1] != canonical[2],
	}, nil
}

func canonicalExistingPath(path string) (string, error) {
	absolute, err := filepath.Abs(path)
	if err != nil {
		return "", err
	}
	canonical, err := filepath.EvalSymlinks(absolute)
	if err != nil {
		return "", fmt.Errorf("canonicalize worktree identity: %w", err)
	}
	if _, err := os.Stat(canonical); err != nil {
		return "", fmt.Errorf("stat worktree identity: %w", err)
	}
	return filepath.Clean(canonical), nil
}

func parseWorktreeList(output []byte) ([]ExistingWorktree, WorktreeBounds, error) {
	records := bytes.Split(output, []byte{0, 0})
	worktrees := make([]ExistingWorktree, 0, min(len(records), maxWorktrees))
	total := 0
	for _, record := range records {
		if len(record) == 0 {
			continue
		}
		candidate := ExistingWorktree{}
		valid, skip := true, false
		for _, raw := range bytes.Split(record, []byte{0}) {
			key, value, found := bytes.Cut(raw, []byte{' '})
			if !found {
				if bytes.Equal(raw, []byte("bare")) || bytes.HasPrefix(raw, []byte("prunable")) {
					skip = true
				}
				continue
			}
			switch string(key) {
			case "worktree":
				candidate.Root = string(value)
			case "HEAD":
				candidate.Head = strings.ToLower(string(value))
			case "branch":
				candidate.Branch = strings.TrimPrefix(string(value), "refs/heads/")
			}
		}
		if candidate.Root == "" || !filepath.IsAbs(candidate.Root) ||
			!validObjectID(candidate.Head) || strings.ContainsAny(candidate.Branch, "\x00\r\n") ||
			len(candidate.Branch) > 512 {
			valid = false
		}
		if skip {
			continue
		}
		if !valid {
			return nil, WorktreeBounds{}, errors.New("malformed Git worktree list")
		}
		canonicalRoot, err := canonicalExistingPath(candidate.Root)
		if err != nil {
			continue
		}
		info, err := os.Stat(canonicalRoot)
		if err != nil || !info.IsDir() {
			continue
		}
		candidate.Root = canonicalRoot
		total++
		if len(worktrees) < maxWorktrees {
			worktrees = append(worktrees, candidate)
		}
	}
	return worktrees, WorktreeBounds{
		Shown: len(worktrees), Total: total, More: max(0, total-len(worktrees)),
	}, nil
}

func validObjectID(value string) bool {
	if len(value) != 40 && len(value) != 64 {
		return false
	}
	for _, character := range value {
		if !strings.ContainsRune("0123456789abcdefABCDEF", character) {
			return false
		}
	}
	return true
}

File: crates/ptrack-git/src/worktree.rs
use std::path::{Path, PathBuf};

use crate::model::{ExistingWorktree, WorktreeBounds, WorktreeIdentity};
use crate::runner::{CancellationToken, RepositoryError, args};
use crate::snapshot::{ExecutionSession, RepositoryService};

const MAX_WORKTREES: usize = 64;

impl RepositoryService {
    /// Validates a selected path as a registered worktree of the project.
    ///
    /// # Errors
    ///
    /// Returns a content-free error when identity, containment, membership,
    /// cancellation, subprocess, resource-bound, or filesystem checks fail.
    pub fn inspect_worktree(
        &self,
        cancellation: &CancellationToken,
        project_root: impl AsRef<Path>,
        candidate: impl AsRef<Path>,
    ) -> Result<WorktreeIdentity, RepositoryError> {
        let project_root = project_root.as_ref();
        let candidate = candidate.as_ref();
        let canonical_candidate =
            canonical_existing_path(candidate, "canonicalize selected worktree path failed")?;
        let mut session = ExecutionSession::new(self.runner(), cancellation);
        let project = inspect_worktree_identity(&mut session, project_root)?;
        let selected = inspect_worktree_identity(&mut session, candidate)?;
        if project.common_git_dir != selected.common_git_dir {
            return invalid("selected worktree belongs to a different repository");
        }
        if !path_within_root(Path::new(&selected.root), &canonical_candidate) {
            return invalid("selected path is outside the inspected worktree");
        }
        let listed_output = session.run(
            project_root,
            &args(["worktree", "list", "--porcelain", "-z"]),
        )?;
        let (listed, _) = parse_worktree_list(&listed_output)?;
        if !listed.iter().any(|worktree| worktree.root == selected.root) {
            return invalid("selected worktree is not registered with the project repository");
        }
        Ok(selected)
    }
}

/// Validates a selected path with a default repository service.
///
/// # Errors
///
/// Returns a content-free error when identity, containment, membership,
/// cancellation, subprocess, resource-bound, or filesystem checks fail.
pub fn inspect_worktree(
    cancellation: &CancellationToken,
    project_root: impl AsRef<Path>,
    candidate: impl AsRef<Path>,
) -> Result<WorktreeIdentity, RepositoryError> {
    RepositoryService::new().inspect_worktree(cancellation, project_root, candidate)
}

fn inspect_worktree_identity(
    session: &mut ExecutionSession<'_>,
    root: &Path,
) -> Result<WorktreeIdentity, RepositoryError> {
    let output = session.run(
        root,
        &args([
            "rev-parse",
            "--path-format=absolute",
            "--is-inside-work-tree",
            "--show-toplevel",
            "--absolute-git-dir",
            "--git-common-dir",
            "--is-bare-repository",
            "--verify",
            "HEAD",
        ]),
    )?;
    let text = std::str::from_utf8(&output)
        .map_err(|_| RepositoryError::InvalidData("malformed worktree identity"))?;
    let lines: Vec<&str> = text.lines().collect();
    if lines.len() != 6 {
        return invalid("malformed worktree identity");
    }
    let inside = lines[0]
        .parse::<bool>()
        .map_err(|_| RepositoryError::InvalidData("selected path is not a non-bare worktree"))?;
    let bare = lines[4]
        .parse::<bool>()
        .map_err(|_| RepositoryError::InvalidData("selected path is not a non-bare worktree"))?;
    if !inside || bare {
        return invalid("selected path is not a non-bare worktree");
    }

    let canonical_root =
        canonical_existing_path(Path::new(lines[1]), "canonicalize worktree identity failed")?;
    let canonical_git_dir =
        canonical_existing_path(Path::new(lines[2]), "canonicalize worktree identity failed")?;
    let canonical_common_git_dir =
        canonical_existing_path(Path::new(lines[3]), "canonicalize worktree identity failed")?;
    for path in [
        &canonical_root,
        &canonical_git_dir,
        &canonical_common_git_dir,
    ] {
        if !std::fs::metadata(path).is_ok_and(|metadata| metadata.is_dir()) {
            return invalid("worktree identity path is not a directory");
        }
    }

    let branch = match session.run(
        &canonical_root,
        &args(["symbolic-ref", "--quiet", "--short", "HEAD"]),
    ) {
        Ok(output) => std::str::from_utf8(&output)
            .map_err(|_| RepositoryError::InvalidData("malformed worktree branch"))?
            .trim()
            .to_owned(),
        Err(RepositoryError::CommandFailed) => String::new(),
        Err(error) => return Err(error),
    };
    if branch.contains(['\0', '\r', '\n']) || branch.len() > 512 {
        return invalid("malformed worktree branch");
    }
    let head = lines[5].trim();
    if !valid_object_id(head) {
        return invalid("malformed worktree HEAD");
    }
    let git_dir = path_string(&canonical_git_dir)?;
    let common_git_dir = path_string(&canonical_common_git_dir)?;
    Ok(WorktreeIdentity {
        root: path_string(&canonical_root)?,
        linked: git_dir != common_git_dir,
        git_dir,
        common_git_dir,
        branch,
        head: head.to_ascii_lowercase(),
    })
}

fn path_within_root(root: &Path, candidate: &Path) -> bool {
    candidate.starts_with(root)
}

fn canonical_existing_path(path: &Path, message: &'static str) -> Result<PathBuf, RepositoryError> {
    let absolute = if path.is_absolute() {
        path.to_owned()
    } else {
        std::env::current_dir()
            .map_err(|_| RepositoryError::Filesystem(message))?
            .join(path)
    };
    let canonical =
        std::fs::canonicalize(absolute).map_err(|_| RepositoryError::Filesystem(message))?;
    std::fs::metadata(&canonical).map_err(|_| RepositoryError::Filesystem(message))?;
    Ok(normalize_canonical(canonical))
}

#[cfg(windows)]
fn normalize_canonical(path: PathBuf) -> PathBuf {
    let value = path.to_string_lossy();
    if let Some(unc) = value.strip_prefix(r"\\?\UNC\") {
        return PathBuf::from(format!(r"\\{unc}"));
    }
    if let Some(local) = value.strip_prefix(r"\\?\") {
        return PathBuf::from(local);
    }
    path
}

#[cfg(not(windows))]
fn normalize_canonical(path: PathBuf) -> PathBuf {
    path
}

fn path_string(path: &Path) -> Result<String, RepositoryError> {
    path.to_str()
        .map(str::to_owned)
        .ok_or(RepositoryError::InvalidData(
            "worktree identity path is not valid UTF-8",
        ))
}

pub(crate) fn parse_worktree_list(
    output: &[u8],
) -> Result<(Vec<ExistingWorktree>, WorktreeBounds), RepositoryError> {
    let mut worktrees = Vec::new();
    let mut total = 0;
    let mut fields = Vec::new();
    for field in output.split(|byte| *byte == 0) {
        if field.is_empty() {
            if !fields.is_empty() {
                parse_worktree_record(&fields, &mut worktrees, &mut total)?;
                fields.clear();
            }
        } else {
            fields.push(field);
        }
    }
    if !fields.is_empty() {
        parse_worktree_record(&fields, &mut worktrees, &mut total)?;
    }
    Ok((
        worktrees,
        WorktreeBounds {
            shown: total.min(MAX_WORKTREES),
            total,
            more: total.saturating_sub(MAX_WORKTREES),
        },
    ))
}

fn parse_worktree_record(
    fields: &[&[u8]],
    worktrees: &mut Vec<ExistingWorktree>,
    total: &mut usize,
) -> Result<(), RepositoryError> {
    let mut candidate = ExistingWorktree::default();
    let mut skip = false;
    for raw in fields {
        let (key, value) = raw
            .iter()
            .position(|byte| *byte == b' ')
            .map_or((*raw, &[][..]), |position| {
                (&raw[..position], &raw[position + 1..])
            });
        match key {
            b"worktree" => {
                utf8(value, "malformed Git worktree list")?.clone_into(&mut candidate.root);
            }
            b"HEAD" => {
                candidate.head = utf8(value, "malformed Git worktree list")?.to_ascii_lowercase();
            }
            b"branch" => {
                let branch = utf8(value, "malformed Git worktree list")?;
                branch
                    .strip_prefix("refs/heads/")
                    .unwrap_or(branch)
                    .clone_into(&mut candidate.branch);
            }
            b"bare" | b"prunable" => skip = true,
            _ => {}
        }
    }
    if skip {
        return Ok(());
    }
    if candidate.root.is_empty()
        || !Path::new(&candidate.root).is_absolute()
        || !valid_object_id(&candidate.head)
        || candidate.branch.contains(['\0', '\r', '\n'])
        || candidate.branch.len() > 512
    {
        return invalid("malformed Git worktree list");
    }
    let Ok(canonical_root) = canonical_existing_path(
        Path::new(&candidate.root),
        "canonicalize worktree identity failed",
    ) else {
        return Ok(());
    };
    if !std::fs::metadata(&canonical_root).is_ok_and(|metadata| metadata.is_dir()) {
        return Ok(());
    }
    candidate.root = path_string(&canonical_root)?;
    *total += 1;
    if worktrees.len() < MAX_WORKTREES {
        worktrees.push(candidate);
    }
    Ok(())
}

fn valid_object_id(value: &str) -> bool {
    matches!(value.len(), 40 | 64) && value.bytes().all(|byte| byte.is_ascii_hexdigit())
}

fn utf8<'a>(input: &'a [u8], message: &'static str) -> Result<&'a str, RepositoryError> {
    std::str::from_utf8(input).map_err(|_| RepositoryError::InvalidData(message))
}

fn invalid<T>(message: &'static str) -> Result<T, RepositoryError> {
    Err(RepositoryError::InvalidData(message))
}
